src/Chapter2.hs

```diff
@@ -727,9 +739,13 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+-- smartReplicate [] = []
+-- smartReplicate (x:xs) = replicate x x ++ smartReplicate xs
+-- smartReplicate xs = foldr (\ x -> (++) (replicate x x)) [] xs
+smartReplicate = foldr (\ x -> (++) (replicate x x)) [] 
```

Comment:
This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).